File: frontend/e2e/dates.ts
// E2E 测试日期基准:统一由「执行当日」推算,不硬编。
//
// 为什么:后端 clock.is_past_slot 以真实时钟判定节次是否已上过。日期一旦成为过去,
// 代课指派被 409 拒绝、销假不再级联——测试会在某个没人动过代码的早晨无声转红
// (原本埋的引信是 2026-11-11)。与后端 tests/dates.py 同一套规则。

// 基准周距今至少 14 天:确保基准周的每一节都还没上过(不受执行时刻影响)。
const LEAD_DAYS = 14

function addDays(day: Date, n: number): Date {
  const out = new Date(day)
  out.setDate(out.getDate() + n)
  return out
}

/** ISO 星期:1=周一 … 7=周日(JS 的 getDay() 是 0=周日)。 */
function isoWeekday(day: Date): number {
  return day.getDay() === 0 ? 7 : day.getDay()
}

/** `day` 当天或之后、最近的指定 ISO 星期。 */
export function onOrAfter(weekday: number, day: Date): Date {
  return addDays(day, (weekday - isoWeekday(day) + 7) % 7)
}

/** yyyy-mm-dd(用本地日期字段,避免 toISOString() 的 UTC 位移把日期倒退一天)。 */
export function iso(day: Date): string {
  const pad = (n: number) => String(n).padStart(2, '0')
  return `${day.getFullYear()}-${pad(day.getMonth() + 1)}-${pad(day.getDate())}`
}

/** 基准周的周一:距今 ≥ LEAD_DAYS,且该周一到「下周三」同月。
 *
 * 同月是硬需求:代课推荐的公平计数与月结统计都以「节次那一天的月份」为范围,
 * 基准周跨月会让「本月已代 N 节」归零。与后端 tests/dates.py 的 base_monday 同规则。
 */
function baseMonday(): Date {
  let mon = onOrAfter(1, addDays(new Date(), LEAD_DAYS))
  for (let i = 0; i < 6; i += 1) {
    if (addDays(mon, 9).getMonth() === mon.getMonth()) return mon
    mon = addDays(mon, 7)
  }
  throw new Error('六周内必有一个「当周到下周三同月」的周一')
}

const monday = baseMonday()
const wednesday = addDays(monday, 2)

export const MON = iso(monday)
export const WED = iso(wednesday)                   // 请假日(多数 spec 的主角)
export const THU = iso(addDays(monday, 3))          // 无请假的对照日
export const FRI = iso(addDays(monday, 4))
export const NEXT_MON = iso(addDays(monday, 7))     // 跨周末的请假结束日
export const WED2 = iso(addDays(monday, 9))         // 下周三(调课补课日/第二张假单)

// 学期起止:包住上面所有日子,前后留缓冲
export const SEM_START = iso(addDays(monday, -30))
export const SEM_END = iso(addDays(monday, 120))

/** 基准周里指定 ISO 星期的那一天(1=周一 … 7=周日)。给「单元格在星期几就请那天的假」用。 */
export function dayOfBaseWeek(weekday: number): string {
  return iso(addDays(monday, weekday - 1))
}

/** 月结统计的查询参数,取「该请假日」所属的年月。
 *
 * 必须用请假日自己算:基准周可能跨月(周一 8/31、周三 9/2),
 * 拿别天的月份去查会查到空月份。
 */
export function statsQuery(day: string): string {
  const [year, month] = day.split('-').map(Number)
  return `&year=${year}&month=${month}`
}

/** 多数 spec 的请假日就是 WED,直接用这个。 */
export const STATS_QUERY = statsQuery(WED)

/** 界面上显示的日期格式:「2026-11-11（星期三）」。 */
const WEEK_LABELS = ['一', '二', '三', '四', '五', '六', '日']
export function withWeekday(day: string): string {
  const [y, m, d] = day.split('-').map(Number)
  return `${day}（星期${WEEK_LABELS[isoWeekday(new Date(y, m - 1, d)) - 1]}）`
}

File: backend/tests/dates.py
"""测试用的日期基准:统一由「执行当日」推算,不硬编。

**为什么不能硬编未来日期**:`clock.is_past_slot` 以真实时钟判定节次是否已上过。
日期一旦成为过去,代课指派会被 409 拒绝、销假不再级联——整套测试在某个没人动过
代码的早晨无声转红(原本埋的引信是 2026-11-11)。

用法:`from tests.dates import SEM_START, SEM_END, WED`。星期采 ISO(1=周一 … 7=周日),
与 `ScheduleEntry.weekday` 及 `leaves.expand` 的 `isoweekday()` 同一套。
"""

from datetime import date, timedelta

# 基准周距今的最小天数。取 14 天的理由:
#   1. 基准周的每一节都必须还没上过——若取「今天」,下午跑测试时上午的节次已过期。
#   2. 有测试会往前一周找日子,前一周也必须仍在未来。
LEAD_DAYS = 14

# 学期起止相对基准周留的缓冲(要能容下 CROSS_WED2 这种往后数周的日子)。
_SEM_LEAD = timedelta(days=30)
_SEM_TAIL = timedelta(days=60)


def on_or_after(weekday: int, day: date) -> date:
    """`day` 当天或之后、最近的指定 ISO 星期(1=周一 … 7=周日)。"""
    if not 1 <= weekday <= 7:
        raise ValueError(f"weekday 需为 1~7(ISO),得到 {weekday}")
    return day + timedelta(days=(weekday - day.isoweekday()) % 7)


def base_monday(today: date | None = None) -> date:
    """基准测试周的周一:距今至少 `LEAD_DAYS` 天,且该周一到「下周三」落在同一个月。

    同月是硬需求,不是美观:代课推荐的公平计数与月结统计都以「受影响节次那一天的月份」
    为范围(`_monthly_sub_counts`)。若 WED 与 WED2 跨月,「林师本月已代 1 节」在 WED2
    那个月会归零,「本月代课少者优先」就验不到。跨月的案例另由 `cross_month_wednesday` 负责。
    """
    mon = on_or_after(1, (today or date.today()) + timedelta(days=LEAD_DAYS))
    for _ in range(6):
        if (mon + timedelta(days=9)).month == mon.month:  # +9 天 = 下周三
            return mon
        mon += timedelta(days=7)
    raise AssertionError("六周内必有一个「当周到下周三同月」的周一")  # pragma: no cover


def cross_month_wednesday(today: date | None = None) -> date:
    """一个周三,且「下一个周三」落在不同月份(供跨月假单拆账测试)。

    任何连续 5 周内必有一个月底的周三,故有限次即可找到。
    """
    wed = on_or_after(3, base_monday(today))
    for _ in range(6):
        if (wed + timedelta(days=7)).month != wed.month:
            return wed
        wed += timedelta(days=7)
    raise AssertionError("六周内必有跨月的周三,推算逻辑有误")  # pragma: no cover


MON = base_monday()
TUE = MON + timedelta(days=1)
WED = MON + timedelta(days=2)
THU = MON + timedelta(days=3)
FRI = MON + timedelta(days=4)
SAT = MON + timedelta(days=5)
SUN = MON + timedelta(days=6)
NEXT_MON = MON + timedelta(days=7)
WED2 = WED + timedelta(days=7)  # 下周三(swap 补课、跨周统计用)

# 跨月的两个周三(11/25 与 12/02 那组硬编日期的动态版)
CROSS_WED = cross_month_wednesday()
CROSS_WED2 = CROSS_WED + timedelta(days=7)

# 学期起止:包住上面所有日子,前后各留缓冲供「学期外」的边界测试
SEM_START = MON - _SEM_LEAD
SEM_END = max(WED2, CROSS_WED2) + _SEM_TAIL
BEFORE_SEM = SEM_START - timedelta(days=1)
AFTER_SEM = SEM_END + timedelta(days=1)
